[LLD][ELF][AArch64] Add support for SHF_AARCH64_PURECODE ELF section flag (3/3)

Il-Capitano opened the pull request:
Add support for the new SHF_AARCH64_PURECODE ELF section flag: https://github.com/ARM-software/abi-aa/pull/304

The general implementation follows the existing one for ARM targets. The output section only has the `SHF_AARCH64_PURECODE` flag set if all input sections have it set.

Related PRs:
* LLVM: https://github.com/llvm/llvm-project/pull/125687
* Clang: https://github.com/llvm/llvm-project/pull/125688

Il-Capitano committed:
[LLD][ELF][AArch64] Add support for SHF_AARCH64_PURECODE ELF section flag (3/3)

Add support for the new SHF_AARCH64_PURECODE ELF section flag: https://github.com/ARM-software/abi-aa/pull/304

The general implementation follows the existing one for ARM targets. The
output section only has the `SHF_AARCH64_PURECODE` flag set if all input
sections have it set.

smithp35 reviewed (commented):
I've concentrated on the lld side. I accidentally left one on the assembler before reading that this change also included https://github.com/llvm/llvm-project/pull/125687 .

Antother possible way of splitting it up is to do:
* llvm-mc (section flag only), llvm-readelf
* lld using llvm-mc for tests
* llvm code-gen from C

However no need to change unless there's further feedback from other reviewers.

Il-Capitano commented on lld/ELF/Driver.cpp:
```diff
@@ -459,6 +459,8 @@ static void checkOptions(Ctx &ctx) {
       ErrAlways(ctx) << "-z gcs-report only supported on AArch64";
     if (ctx.arg.zGcs != GcsPolicy::Implicit)
       ErrAlways(ctx) << "-z gcs only supported on AArch64";
+    if (ctx.arg.zExecuteOnlyReport != "none")
```
No particular reason other than this patch is concerned with only `SHF_AARCH64_PURECODE`, and I hadn't thought about implementing this for ARM as well. I could add support for it here if you prefer, or open another PR implementing that once this gets merged.

smithp35 replied:
When I first saw the PR, my immediate thought was to separate -z execute-only-report into a separate patch and implement for Arm/AArch64. This may be worth doing as you can write tests for that in that patch. Will also help speed up the LLD review for this part as there's less code to look at.

Il-Capitano replied:
That sounds good to me. I've already written the test cases in the mean time, so I'll wait for MaskRay to chime in if I should remove that part from this patch.

Il-Capitano commented on lld/ELF/OutputSections.cpp:
```diff
@@ -42,7 +42,8 @@ using namespace lld::elf;
 uint32_t OutputSection::getPhdrFlags() const {
   uint32_t ret = 0;
-  if (ctx.arg.emachine != EM_ARM || !(flags & SHF_ARM_PURECODE))
+  if ((ctx.arg.emachine != EM_ARM || !(flags & SHF_ARM_PURECODE)) &&
```
I wanted to separate out `SHF_ARM_*` and `SHF_AARCH64_*` in the code in order to not confuse the two architectures.
Adding `static_assert(SHF_ARM_PURECODE == SHF_AARCH64_PURECODE)`, and using only one of the flags in the condition could also work.

Other than the condition being more verbose like this being a downside, I think I have a weak preference for writing out the two flags explicitly, and not mixing different architectures.

smithp35 replied:
That's fine with me. I've added MaskRay (lld maintainer) to the review. If he has a strong opinion then we'll go with his preference.

MaskRay replied:
Perhaps something  like
```cpp
bool purecode = (emachine == EM_ARM && (flags & SHF_ARM_PURECODE))) || (emachine == EM_AARCH && (flags & SHF...));
if (!purecode)
  ret |= PF_R;
```

Il-Capitano replied:
That looks good to me.

Il-Capitano commented on lld/ELF/Driver.cpp:
```diff
@@ -2844,6 +2848,40 @@ static void readSecurityNotes(Ctx &ctx) {
+  auto reportUnless = [&](StringRef config, bool cond) -> ELFSyncStream {
+    if (cond)
+      return {ctx, DiagLevel::None};
+    return report(config);
```
Good catch. I copied the structure from the other `-z *-report` checks above, where `report` is used elsewhere in that function. I'll remove it.

Il-Capitano commented on llvm/include/llvm/BinaryFormat/ELF.h:
```diff
@@ -1300,7 +1300,10 @@ enum : unsigned {
-  SHF_ARM_PURECODE = 0x20000000
+  SHF_ARM_PURECODE = 0x20000000,
+
+  // Section contains only program instructions and no program data.
```
I made a comment on the PR concerned with the LLVM part about this: https://github.com/llvm/llvm-project/pull/125687/files#r1941424500

Il-Capitano committed:
Address some review feedback

Il-Capitano committed:
Add tests for -z execute-only-report

Il-Capitano commented on lld/test/ELF/aarch64-execute-only.s:
```diff
@@ -0,0 +1,33 @@
+// REQUIRES: aarch64
```
Done.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -1557,10 +1559,12 @@ static void readConfigs(Ctx &ctx, opt::InputArgList &args) {
+      std::make_pair("cet-report", &ctx.arg.zCetReport),
+      std::make_pair("gcs-report", &ctx.arg.zGcsReport),
+      std::make_pair("pauth-report", &ctx.arg.zPauthReport),
+      std::make_pair("execute-only-report", &ctx.arg.zExecuteOnlyReport)};
```
sort alphabetically

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -2844,6 +2848,36 @@ static void readSecurityNotes(Ctx &ctx) {
+    return {ctx, DiagLevel::None};
+  };
+
+  for (ELFFileBase *file : ctx.objectFiles) {
```
likely faster by iteration over executable `OutputSection`s and check `InputSectionDescription`s and then `InputSection`s.

Il-Capitano replied:
This check will have to be relocated into `Writer` then, I'll do that.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -2844,6 +2848,36 @@ static void readSecurityNotes(Ctx &ctx) {
+  };
+
+  for (ELFFileBase *file : ctx.objectFiles) {
+    for (InputSectionBase *section : file->getSections()) {
```
For new code, use `osec` for OutputSection and `sec` for InputSectionBase/InputSection.

MaskRay commented on lld/ELF/Driver.cpp:
```diff
@@ -2844,6 +2848,36 @@ static void readSecurityNotes(Ctx &ctx) {
+      if (outputSection && outputSection->name == ".text") {
+        reportUnless(ctx.arg.zExecuteOnlyReport,
+                     section->flags & SHF_AARCH64_PURECODE)
+            << file << ": -z execute-only-report: section " << section->name
```
We have `operator<<(const ELFSyncStream &s, const InputSectionBase *sec)` to describe both the file name and the section name.

Perhaps like this: `"-z execute-only-report:" << sec << " does not have ..."`

Il-Capitano replied:
I didn't know about that, thanks for pointing it out.

MaskRay reviewed (commented):
The lld part should go to a separate patch

MaskRay commented on llvm/test/CodeGen/AArch64/execute-only-section.ll:
```diff
@@ -0,0 +1,21 @@
+; RUN: llc -mtriple=aarch64 -mattr=+execute-only %s -o - | FileCheck %s
+
+; CHECK:     .section .text,"axy",@progbits,unique,0
```
Should test more cases:

* a COMDAT text section
* a rodata section doesn't get `y`

MaskRay commented on llvm/test/MC/ELF/AArch64/execute-only-section.s:
```diff
@@ -0,0 +1,43 @@
+// RUN: llvm-mc -filetype=obj -triple aarch64-unknown-linux-gnu %s -o - \
+// RUN: | llvm-readobj -S --symbols - | FileCheck %s
```
also need a `llvm-readelf -S` test

MaskRay commented on llvm/test/MC/ELF/AArch64/execute-only-section.s:
```diff
@@ -0,0 +1,43 @@
+        .size   foo, .Lfunc_end0-foo
+        .cfi_endproc
+
+        .section        ".note.GNU-stack","",@progbits
```
delete

jh7370 commented:
> The lld part should go to a separate patch

This is the separate patch, but the previous commits from other PRs are included in it because Github is terrible...

Il-Capitano committed:
Address review feedback from MaskRay

Il-Capitano committed:
Move -z execute-only-report checking into Writer.cpp

Il-Capitano commented:
@MaskRay I've addressed your review comments on the LLVM part of the patch in https://github.com/llvm/llvm-project/pull/125687. I won't update the code here in order to not clutter up this PR with unnecessary commits.

Il-Capitano committed:
Resolve merge conflicts after rebase

Il-Capitano commented:
I've rebased this patch onto main, since https://github.com/llvm/llvm-project/pull/125687 has been merged. Only the LLD part is included now.

smithp35 commented on lld/ELF/Writer.cpp:
```diff
@@ -2179,6 +2181,34 @@ template <class ELFT> void Writer<ELFT>::checkExecuteOnly() {
                             "data and code";
 }
 
+// Check that all input sections of .text have the SHF_AARCH64_PURECODE section
```
While I think this will work in a lot of cases, there can be multiple executable output sections covered by a PF_E program header.  It is true that the output of executable sections from the compiler will use a .text prefix and end up in the .text output section. However there can be named sections, and assembler sections that may not follow that convention. These will get placed as orphans and will get missed by this check.

What I think needs to happen is to find all OutputSections that are mapped to a program header that has the execute `PF_E` and read `PF_R` flag. There is no point in checking OutputSections that are mapped to a program header with `PF_E` and not `PF_R` as we already know all those sections are execute-only by construction.

When there's more than one program header with `PF_E` and `PF_R` it will be useful to logically group the output per program header. While in Android there will only be one executable program header, in bare-metal systems there can be more than one. It is also likely that in an embedded system some program headers are supposed to be execute-only and some not. In particular when a toolchain only has non-execute-only C and C++ libraries these need to be separated into their own program header.

Il-Capitano replied:
My thinking was that anything outside of .text is the responsibility of the toolchain, but I hadn't thought about other use cases you mentioned. Thanks for pointing those out.

> When there's more than one program header with PF_E and PF_R it will be useful to logically group the output per program header.

Can you clarify this a bit? With the current code if I check every `PF_R | PF_X` section I get this output currently from one of my test cases:
```
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/Scrt1.o:(.text) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/crti.o:(.text) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /usr/lib/gcc/aarch64-linux-gnu/12/crtbeginS.o:(.text) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /usr/lib/gcc/aarch64-linux-gnu/12/crtendS.o:(.text) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/crtn.o:(.text) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: test.lto.o:(.text) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: test.lto.o:(.text.b) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/crti.o:(.init) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/crtn.o:(.init) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/crti.o:(.fini) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: /lib/aarch64-linux-gnu/crtn.o:(.fini) does not have SHF_AARCH64_PURECODE flag set
ld.lld: warning: -z execute-only-report: <internal>:(.plt) does not have SHF_AARCH64_PURECODE flag set
```
The diagnostics follow each other by their output section, `.text`, `.init`, `.fini`, and `.plt`. (Also, the last warning has `<internal>:(.plt)`, which I'm not sure how to handle yet.)

smithp35 replied:
The outputs from the c runtime make sense as I expect no-one has added SHF_AARCH64_PURECODE to these files. I'm surprised that these didn't get picked up by your earlier search through the .text OutputSection though.

The `.plt` is a synthetic linker generated section that reads data from the `.plt.got` section (which should be in read-write data). Another example of linker generated executable sections are the long-branch extension thunks. Of these the position independent ADRP thunk is execute only, but the absolute non-pic one is not.

I can think of a couple of ways to handle this:
* Add the SHF_AARCH64_PURECODE flag to the .plt on AArch64 (and Arm platforms). Add the SHF_AARCH64_PURECODE flag to the thunks that are execute-only. Can do the same on ARM.
* Selectively filter out the synthetic sections. For example you can do that for the .plt. The thunks are a bit riskier as some may not be execute-only. The Arm thunks have some execute-only specific thunks.

We may have to add an XO variant of the AARCH64ABSLongThunk https://github.com/llvm/llvm-project/blob/main/lld/ELF/Thunks.cpp#L637 that uses MOVZ/MOVK to synthesize the immediate. Or just use the ADRP thunk when XO is being used.

There's a precedent for this with v6-m which adds a somewhat horrific XO thunk (v6-m lacks MOVT and MOVW) https://github.com/llvm/llvm-project/blob/main/lld/ELF/Thunks.cpp#L937

Il-Capitano replied:
> The outputs from the c runtime make sense as I expect no-one has added SHF_AARCH64_PURECODE to these files. I'm surprised that these didn't get picked up by your earlier search through the .text OutputSection though.

They were there all along in my local testing, I just ignored them because it's expected that they don't have the SHF_AARCH64_PURECODE flag. I was able to get CRT files with the flag set by messing around with `llvm-objcopy`, and I could produce a final executable with PF_R stripped.


> I can think of a couple of ways to handle this:
> 
> * Add the SHF_AARCH64_PURECODE flag to the .plt on AArch64 (and Arm platforms). Add the SHF_AARCH64_PURECODE flag to the thunks that are execute-only. Can do the same on ARM.
> * Selectively filter out the synthetic sections. For example you can do that for the .plt. The thunks are a bit riskier as some may not be execute-only. The Arm thunks have some execute-only specific thunks.

Adding the purecode flag for `.plt` sections seems like the better approach to me. I'm happy to do a separate patch that addresses this for both ARM and AArch64.

As for the function call thunks, I'll add your suggestion of handling them to this patch, since it's AArch64-specific and closely tied to the new section flag.

Thank you for the helpful links!

smithp35 reviewed (commented):
Personally I think it would be better to separate out the execute-only-report into a separate patch. I think the basic support is already there and could be committed. We could also generalise the report to include ARM if it were committed separately.

Il-Capitano commented:
> Personally I think it would be better to separate out the execute-only-report into a separate patch. I think the basic support is already there and could be committed. We could also generalise the report to include ARM if it were committed separately.

OK, I'll do that.

Il-Capitano committed:
Remove `-z execute-only-report`

Il-Capitano commented:
The code formatting check fail seems to be unrelated to this patch. `clang-format` is happy on my end.

Il-Capitano committed:
Add execute-only compatible thunks

smithp35 commented on lld/ELF/Thunks.cpp:
```diff
@@ -1482,15 +1522,20 @@ Thunk::Thunk(Ctx &ctx, Symbol &d, int64_t a)
+  bool isPureCode = sec.getParent()->flags & SHF_AARCH64_PURECODE;
   if (ctx.arg.picThunk)
     return std::make_unique<AArch64ADRPThunk>(ctx, s, a, mayNeedLandingPad);
+  if (isPureCode)
```
LLD tends not to use temporary variables that are used once. I suggest:
```
  if (sec.getParent()->flags & SHF_AARCH64_PURECODE)
```

smithp35 commented on lld/test/ELF/Inputs/big-execute-only.s:
```diff
@@ -0,0 +1,3 @@
+.section .text,"axy",@progbits,unique,0
```
I don't think you need this file. We aren't using the .text section as big is an absolute symbol.

You should be able to use the lld command line option --defsym big=0x1111222233334444 

There's an example in aarch64-thunk-bti.s

As an aside, if you do need more than one file, new tests to LLD are preferring to use split-file to have all inputs in one file. Example in aarch64-thunk-bti.s

Il-Capitano replied:
I was basing the tests off of aarch64-call26-thunk.s. Didn't know about `--defsym`, and it's definitely the better approach. Thanks!

smithp35 commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,22 @@
+
+// CHECK: Disassembly of section .text:
+// CHECK-EMPTY:
+// CHECK-NEXT: <_start>:
```
Can use `CHECK-LABEL:` here.

smithp35 commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,22 @@
+// CHECK-NEXT:    210120:       bl      0x210124
+// CHECK:      <__AArch64AbsXOLongThunk_big>:
+// CHECK-NEXT:    210124:       mov  x16, #0x4444
+// CHECK-NEXT:    210128:       movk x16, #0x3333, lsl #16
```
lld test checks prefer to only output the first instruction address of a sequence. So we'd keep 210124, but not 210128-210134.

smithp35 commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,22 @@
+// CHECK-EMPTY:
+// CHECK-NEXT: <_start>:
+// CHECK-NEXT:    210120:       bl      0x210124
+// CHECK:      <__AArch64AbsXOLongThunk_big>:
```
Can use CHECK-LABEL:

smithp35 commented on lld/test/ELF/aarch64-jump26-thunk-execute-only.s:
```diff
@@ -0,0 +1,22 @@
+_start:
+    b big
+
+// CHECK: Disassembly of section .text:
```
Similar comments to before. Can use CHECK-LABEL and remove all but the first address in a sequence.

smithp35 reviewed (commented):
Thanks for the update. I've checked the encodings and relocations for the Thunk and that all looks good. I've made some suggestions for the test, given some lld test conventions.

Il-Capitano committed:
Address review feedback

smithp35 reviewed (approved):
Thanks for the update. LGTM on my side. Please give the maintainer @MaskRay some time to comment before merging.

MaskRay commented on lld/ELF/ScriptParser.cpp:
```diff
@@ -1411,6 +1411,7 @@ static std::optional<uint64_t> parseFlag(StringRef tok) {
       .Case(CASE_ENT(SHF_COMPRESSED))
       .Case(CASE_ENT(SHF_EXCLUDE))
       .Case(CASE_ENT(SHF_ARM_PURECODE))
+      .Case(CASE_ENT(SHF_AARCH64_PURECODE))
```
Update the test test/ELF/input-section-flags.s to test SHF_AARCH64_PURECODE

(rebase. I am adding a SHF_ARM_PURECODE test there.)

MaskRay commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,21 @@
+// CHECK-NEXT:             movk x16, #0x2222, lsl #32
+// CHECK-NEXT:             movk x16, #0x1111, lsl #48
+// CHECK-NEXT:             br   x16
+
```
delete blank lines at the end

MaskRay commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,21 @@
+// CHECK: Disassembly of section .text:
+// CHECK-EMPTY:
+// CHECK-LABEL: <_start>:
+// CHECK-NEXT:     210120: bl      0x210124
```
omit the insignificant address and `bl {{.*}} <__AArch64AbsXOLongThunk_big>` so that the test will not need an update when address layout changes

MaskRay commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,21 @@
+// CHECK-LABEL: <_start>:
+// CHECK-NEXT:     210120: bl      0x210124
+// CHECK-LABEL: <__AArch64AbsXOLongThunk_big>:
+// CHECK-NEXT:     210124: mov  x16, #0x4444
```
drop address

MaskRay commented on lld/test/ELF/aarch64-execute-only.s:
```diff
@@ -0,0 +1,33 @@
+// RUN: ld.lld %t.o -o %t.so -shared
+// RUN: llvm-readelf -l %t.so | FileCheck --implicit-check-not=LOAD %s
+
+// RUN: echo ".section .foo,\"ax\"; \
```
unneeded line wrap

MaskRay commented on lld/test/ELF/aarch64-execute-only.s:
```diff
@@ -0,0 +1,33 @@
+// REQUIRES: aarch64
+
+// RUN: llvm-mc -filetype=obj -triple=aarch64-linux-none %s -o %t.o
```
The triple aarch64-linux-none does not make sense. Just `aarch64`

(While you could use `aarch64-linux`, `aarch64` (unknown ELF) makes it clear that the test doesn't use Linux specific features)

MaskRay commented on lld/test/ELF/aarch64-jump26-thunk-execute-only.s:
```diff
@@ -0,0 +1,21 @@
+// CHECK-NEXT:             movk    x16, #0x2222, lsl #32
+// CHECK-NEXT:             movk    x16, #0x1111, lsl #48
+// CHECK-NEXT:             br      x16
+
```
delete blank line at file end

Il-Capitano committed:
Address review feedback

Il-Capitano commented on lld/test/ELF/input-section-flags.s:
```diff
@@ -15,7 +15,8 @@
 # RUN: .outsec5 : { INPUT_SECTION_FLAGS(SHF_STRINGS) *(.sec.*) } \
-# RUN: .outsec6 : { INPUT_SECTION_FLAGS(!SHF_TLS & !SHF_EXCLUDE & !SHF_COMPRESSED & !SHF_ARM_PURECODE) *(.sec.*) } \
+# RUN: .outsec6 : { INPUT_SECTION_FLAGS(!SHF_TLS & !SHF_EXCLUDE & !SHF_COMPRESSED & \
+# RUN:                                  !SHF_ARM_PURECODE & !SHF_AARCH64_PURECODE) *(.sec.*) } \
```
@MaskRay Is this what you had in mind for this test? IIUC this line tests if we can parse these flags in linker scripts, since they aren't tested elsewhere. But I can't find similar tests for `SHF_GROUP`, `SHF_OS_NONCONFORMING`, `SHF_LINK_ORDER`, or `SHF_INFO_LINK`, so I'm not sure what the intent is.

MaskRay replied:
Yes, just add SHF_AARCH64_PURECODE here, which demonstrates that the arch-specific flag is actually parsed for all targets.

Arguably the test is not perfect: we don't test that the flag value is expected, but for the purpose of ensuring that this flag is parsed, this test is ok.

(Right that many generic flags are not tested, but they are usually never used in linker scripts)

Il-Capitano replied:
Cool, thanks for the help!

Il-Capitano commented:
FYI I don't have commit access, so you'll have to merge this PR for me.

MaskRay committed:
improve test

MaskRay commented on lld/ELF/Thunks.cpp:
```diff
@@ -663,6 +676,33 @@ void AArch64ABSLongThunk::addLongMapSyms() {
+  // If mayNeedLandingPad is true then destination is an
+  // AArch64BTILandingPadThunk that defines landingPad.
+  assert(!mayNeedLandingPad || landingPad != nullptr);
+  uint64_t s = mayNeedLandingPad
```
You copied this piece of intricate code. I think we need a test, perhaps a simplified aarch64-thunk-bti.s #108989

Otherwise, it is fairly easy to cause regression here when someone refactors this code.

Il-Capitano replied:
Good point, I'll be able to work on this tomorrow.

Il-Capitano replied:
Done.

Il-Capitano committed:
Add test for a BTI landing pad target in execute-only thunks

smithp35 reviewed (approved):
Test case looks good to me. I think the generic parts of the BTI insertion are already tested in aarch64-thunk-bti.s , leaving the execute-only specific parts here.

As you don't have permission to merge, I'll leave that for MaskRay once he's happy with the changes.

MaskRay commented on lld/test/ELF/aarch64-call26-thunk-execute-only.s:
```diff
@@ -0,0 +1,20 @@
+.section .text,"axy",@progbits,unique,0
+.globl _start
+_start:
+    bl big
```
This is so similar to to aarch64-jump26-thunk-execute-only.s that they can be merged. We try to combine similar tests as many times having fewer files is easier to read.

```
bl big
b big
```

Il-Capitano replied:
Done.

Il-Capitano committed:
Merge aarch64-call26-thunk-execute-only.s and aarch64-jump26-thunk-execute-only.s
